java-squid/src/main/java/org/sonar/java/SonarComponents.java

```diff
@@ -183,30 +188,77 @@ public RuleKey getRuleKey(JavaCheck check) {
-          }
-        } else {
-          issueBuilder.effortToFix(cost);
-        }
-        Issue issue = issueBuilder.build();
-        issuable.addIssue(issue);
+    if (key == null) {
+      return;
+    }
+    File file = analyzerMessage.getFile();
+    InputPath inputPath = inputPathFromIOFile(file);
+    if (inputPath == null) {
+      return;
+    }
+    Double cost = analyzerMessage.getCost();
+    assertEffortToFixIsNotRequired(check, cost);
+    if (IS_SONARQUBE_52) {
+      reportIssueAfterSQ52(analyzerMessage, key, inputPath, cost);
+    } else {
+      reportIssueBeforeSQ52(inputPath, key, cost, analyzerMessage.getMessage(), analyzerMessage.getLine());
+    }
+  }
+
+  @VisibleForTesting
+  void reportIssueAfterSQ52(AnalyzerMessage analyzerMessage, RuleKey key, InputPath inputPath, Double cost) {
+    CompatIssue issue = CompatIssue.create(context, key, cost);
+    AnalyzerMessage.TextSpan textSpan = analyzerMessage.primaryLocation();
+    if (textSpan == null) {
+      // either an issue at file or folder level
+      issue.setPrimaryLocationOnFile(inputPath, analyzerMessage.getMessage());
+    } else {
+      issue.setPrimaryLocation((InputFile) inputPath, analyzerMessage.getMessage(), textSpan.startLine, textSpan.startCharacter, textSpan.endLine, textSpan.endCharacter);
+    }
+    for (AnalyzerMessage location : analyzerMessage.secondaryLocations) {
+      AnalyzerMessage.TextSpan secondarySpan = location.primaryLocation();
+      issue.addSecondaryLocation(
+        inputFromIOFile(location.getFile()), secondarySpan.startLine, secondarySpan.startCharacter, secondarySpan.endLine, secondarySpan.endCharacter, location.getMessage());
+    }
+    issue.save();
+  }
+
+  private void reportIssueBeforeSQ52(InputPath inputFile, RuleKey key, @Nullable Double cost, String message, @Nullable Integer line) {
+    Issuable issuable = issuableFor(inputFile);
+    if (issuable != null) {
+      Issuable.IssueBuilder newIssueBuilder = issuable.newIssueBuilder()
+        .ruleKey(key)
+        .message(message)
+        .line(line)
+        .effortToFix(cost);
+      issuable.addIssue(newIssueBuilder.build());
+    }
+  }
+
+  private static void assertEffortToFixIsNotRequired(JavaCheck check, @Nullable Double cost) {
+    if (cost == null) {
+      Annotation linear = AnnotationUtils.getAnnotation(check, SqaleLinearRemediation.class);
+      Annotation linearWithOffset = AnnotationUtils.getAnnotation(check, SqaleLinearWithOffsetRemediation.class);
+      if (linear != null || linearWithOffset != null) {
+        throw new IllegalStateException("A check annotated with a linear sqale function should provide an effort to fix");
       }
     }
   }
+
+  private static boolean isSonarQube52() {
+    try {
+      Issuable.IssueBuilder.class.getMethod("newLocation");
+      return true;
+    } catch (NoSuchMethodException e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)